### DISCUSSION TITLE
Custom user attribute from the registration form does not get saved

### DISCUSSION THREAD

**Discussion by hodossy** [keycloak/keycloak#13549] [Q&A] [answered]
I have a custom theme with a custom `register.ftl`. I just copied the one from the `base` theme, and added a new field:

```freemaker
            <div class="${properties.kcFormGroupClass!}">
                <div class="${properties.kcLabelWrapperClass!}">
                    <label for="user.attributes.company" class="${properties.kcLabelClass!}">${msg("company")}</label>
                </div>
                <div class="${properties.kcInputWrapperClass!}">
                    <input type="text" id="user.attributes.company" class="${properties.kcInputClass!}" name="company"
                           value="${(register.formData['user.attributes.company']!'')}" autocomplete="company" required
                           aria-invalid="<#if messagesPerField.existsError('user.attributes.company')>true</#if>"
                    />

                    <#if messagesPerField.existsError('user.attributes.company')>
                        <span id="input-error-user.attributes.company" class="${properties.kcInputErrorMessageClass!}" aria-live="polite">
                            ${kcSanitize(messagesPerField.get('user.attributes.company'))?no_esc}
                        </span>
                    </#if>
                </div>
            </div>
```

I was following [this guide](https://www.keycloak.org/docs/latest/server_development/#registration-page), but after registering a new user, the user attributes page is empty, even though I have defined a company. Is a custom java SPI required for this? Any ideas what am I doing wrong here? Thanks for the help!

**Comment by sventorben** [comment, marked as answer]
I think the `name` attribute of you `input` field has to be `user.attributes.company` and not only `company`.

**Reply by hodossy** [reply]
Thanks, nice catch! I thought that the `id` field is the important one. After changing the `name` attribute as suggested, the field gets saved!

**Comment by martinjinda** [comment]
hey @hodossy does your js validation works correctly?

**Comment by abebehailu** [comment]
Hi @hodossy, @sventorben , @martinjinda,

I am having issue with having this simple example to include validation of the attribute. This only gives frontend validation. The * messagesPerField.existsError('user.attributes.company')* does not seem to kick in. 

Do any one of you know how to enable a field to be required with backend validation?

/Thanks

**Reply by martinjinda** [reply]
I ended up adding to all inputs the HTML attribute "required", which half solved the problem but still would love to have the frontend validation as well.

**Reply by abebehailu** [reply]
@hodossy  Ok, the "required" HTML attribute gives client side validation. You mean you would also want to have backend validation? At least, that is what I am after.

The Server Developer Guide has an example but it does not include validation at all https://www.keycloak.org/docs/latest/server_development/#registration-page

@sventorben @hodossy do you guys have som input here? Would be much appreciated!

/Thanks

**Reply by sventorben** [reply]
> @sventorben @hodossy do you guys have som input here? Would be much appreciated!
> 

Have you tried the declarative user profile? It has validation built in.

https://www.keycloak.org/docs/latest/server_admin/#user-profile

**Reply by abebehailu** [reply]
@sventorben Thanks for the response. Yes I have! It has great features. However, it is still in Technology Preview which is not recommended to use in production.

/Thanks

**Reply by amelongo** [reply]
Assuming 'city' is the custom attribute, here's how the declarative attribute and its error message are implemented in 'register.ftl' **You don't need** to create an explicit Register SPI for that:

         ```
 ```
        <div class="input-group mb-3">
                <span class="input-group-text"><i class="icon cil-factory"></i></span>
                <input id="profile.attributes.city" name="profile.attributes.city" value="${(register.formData['profile.attributes.city']!'')}"
                       class="form-control <#if messagesPerField.existsError('city')>is-invalid</#if>"
                       placeholder="${msg("profile.attributes.city")}" autocomplete="city" type="text"/>
                <#if messagesPerField.existsError('city')>
                    <div id="input-error-city" class="invalid-feedback">
                        ${kcSanitize(messagesPerField.getFirstError('city'))?no_esc}
                    </div>
                </#if>
            </div>
```
```

**Comment by Sieboldianus** [comment]
Is there a way to also achive the same with the declarative user profile? (25.0.6)

Atm. all declarative profile fields are reset on validation error, and the user has to input all again. I tried changing register.ftl, but this means manually overriding the whole declarative user profile config.

I have 25 fields the user needs to enter on registration. It is very frustrating having to retype all these fields on every validation error.
